### ISSUE TITLE
Abnormal memory consumption of Jacobian phase in the NB

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#10230]
I recently added [HarmonicOscillator_N_819200](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_819200.err) to the [LargeTestSuite_NB](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/LargeTestSuite_NB.html) job.

The system has 1.6 million variables, which are all states. As the system is non-stiff, the explicit dopri45 solver is currently used to simulate it, so the Jacobian is actually not needed at all - it may be worth considering optionally disabling Jacobian code generation in the NB, but that's another story. Unfortunately the test [fails](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_819200.err), I guess because the allocated 60 GB are not enough.

Looking at the two smaller cases, [HarmonicOscillator_N_204800](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_204800.err) allocates almost 9 GB of memory for the Jacobian phase, and [HarmonicOscillator_N_409600](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_409600.err) almost 18 GB. The scaling is not superlinear, which is good, but the code is probably quite inefficient memory-wise. 

Consider that N = 409600 case has 1.6 M non-zero entries in the Jacobian, each of them consumes about 10 kB of RAM. I'm not sure why, but I guess this amount of memory could probably be reduced by at least one order of magnitude without too much effort. Maybe more. This is essential to break the 1M state limit and head towards the 10M goal.

**Comment by casella** [comment]
See @adrpo's comment in https://github.com/OpenModelica/OpenModelica/issues/10400#issuecomment-1469724200

**Comment by casella** [comment]
I'm not sure if @perost or @adrpo has done something in the meantime, but now we can run [HarmonicOscillator_N_819200](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_819200.err) successfully, despite the 35 GB allocated for the Jacobian generation. In fact, SimCode allocates another 50 GB, bringing the total allocation to 110 GB. Despite that, and despite the fact that the task is only allocated 62 GB, the test runs succesfully.

Apparently something happened in-between that may have improved the capability of the garbage collector to recover some memory.

I can try to run the N_1638400 model and see what happens.

**Comment by casella** [comment]
[HarmonicOscillator_N_819200](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_819200.err) has stopped working again some time ago.

When building the largest working model, [Mechanical.HarmonicOscillator.HarmonicOscillator_N_409600.](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_409600.err), which has 819200 equations, the new backend reports:
```
Notification: Performance of Causalize: time 5.491/5.494, allocations: 3.348 GB / 3.972 GB, free: 160.7 MB / 1.046 GB
...
Notification: Performance of Causalize: time 16.49/21.99, allocations: 10.41 GB / 14.38 GB, free: 0.8733 GB / 2.429 GB
...
Notification: Performance of Jacobian: time 41.96/63.97, allocations: 18.25 GB / 32.63 GB, free: 385.9 MB / 3.427 GB
```
The model has a really simple structure:
```modelica
      for i in 1:N loop
        der(x[i]) = v[i];
      end for;
      m*der(v[1]) = k*(x[2]-x[1]);
      for i in 2:N - 1 loop
        m*der(v[i]) = k*(x[i-1] - x[i]) + k*(x[i+1] - x[i]);
      end for;
      m*der(v[N]) = k*(x[N-1]-x[N]);
    initial equation
      x[1] = N;
      v[1] = 0;
      for i in 2:N loop
        x[i] = 0;
        v[i] = 0;
      end for;
```
The first causalization involves 819200 equations, with each equation involving at most three variables, and allocates about 3 GB of memory. Hence, the expanded E-V graph has about 1 million E nodes, 1 million V nodes, and a few million arcs, with each of these basic objects consuming somthing of order of 1 kB of memory. This seems really a lot to me. All the information pertaining the variables (in particular the attributes) should have already been allocated before starting causalization and Jacobian evaluation, so as I understand, a parsimonious representation of the E-V graph would only require a few bytes per arc.

Ditto for the Jacobian, which I expect to have about 2 M nonzero element. 16 GB means 8 kB allocated for each non-zero element, which is way too much.

@mahge, @perost, could you run some profiling on this test case and figure out what are function calls the consume the most memory?

Thanks!

**Comment by casella** [comment]
In the meantime the situation has slightly improved: now [LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_409600](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_409600.err) reports:

```
Notification: Performance of Jacobian: time 26.9/54.77, allocations: 10.51 GB / 25.14 GB, free: 301.3 MB / 2.93 GB
```
memory allocation and time almost halved.

PR #13019 may actually fix this problem for good, let's see what happens when it's merged in.

**Comment by phannebohm** [comment]
#13019 mostly improved models with large algebraic loops. I'm not sure this is the same problem, but let's see.

**Comment by casella** [comment]
It is always difficult to figure out which "Jacobians" the Jacobian phase refers to 😅 

Is it the Jacobians of the strong components or the Jacobian of the ODE system (DAE when --daeMode is set)?

**Comment by phannebohm** [comment]
Sorry, I wasn't clear. I meant it's about whether there is a strong component which is a large algebraic loop present in the system, no matter which Jacobian.

From what I can see, there is no algebraic loop at all in the model, so there is probably still another bottle neck.

**Comment by casella** [comment]
No major changes to [LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_409600](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/files/LargeTestSuite_NB_LargeTestSuite.Mechanical.HarmonicOscillator.HarmonicOscillator_N_409600.err) performance after #13019 
```
Notification: Performance of Jacobian: time 28.54/56.32, allocations: 12.4 GB / 27.04 GB, free: 8.457 MB / 2.909 GB
```
@perost, if you do some profiling, can you see some obvious low-hanging fruit to improve the performance in the Jacobian and Causalize phases of the new backend?

**Comment by casella** [comment]
After #10232, we are no longer generating the Jacobian for these models, and just using an explicit solver. Most of the backend time for the HarmonicOscillator models in [LargeTestSuite_NB](https://libraries.openmodelica.org/branches/heavy_tests/LargeTestSuite_NB/LargeTestSuite_NB.html) is spent doing causalization, which is still scalar, and it's O(N), which is OK.

So, we can consider this issue to be resolved.
